### PR TITLE
Create MapSet[T] and ImmutableMapSet[T] types and change Entity.Parents to be a ImmutableMapSet[EntityUID]

### PR DESCRIPTION
**Pull request by patjakdev**
*Issue #, if available:* None

*Description of changes:*

This PR adds a generic `MapSet[T]` and `ImmutableMapSet[T]` for all the places we want the semantics of a set but are annoyed by the inconvenience of using the `map[T]struct{}` idiom in Go.

In particular, the parents of an `Entity` have set semantics. With the current slice-based implementation, time complexity of the evaluation of `in` is linear. Utilizing a `ImmutableMapSet[EntityUID]` instead gives us constant time complexity on average, which is better. That being said, for small sets, that constant time is actually slightly greater than the time it takes to perform a linear search. I will consider addressing that in a future PR, but for now I'm happy to just have set semantics for `Entity` parents.

The `ImmutableMapSet[EntityUID]` is projected publicly through the `types` package as `EntityUIDSet`, with appropriate constructors also projected. For now, at least, I thought it prudent not to allow external users to import and use `MapSet[T]` directly.

I've also replaced any existing internal usages of `map[T]struct{}` in the module with `MapSet[T]` or `ImmutableMapSet[T]` as appropriate.

Important notes:
1. `MapSet[T]` has no ordering guarantees. This applies to `MarshalJSON` as well, meaning there's no predictable JSON encoding for a given `Set`. Instead of using `comparable`, we could invent an `Item` interface that embeds `comparable` and also has a `Less()` method, but then this set couldn't easily be used for native types like `string`, which feels like a pretty lame drawback.
2. While `MapSet` is implemented in `internal` and therefore can't be directly imported by external users, the `Entity.Parents` is now of type `MapSet[EntityUID]` and therefore changes to the `MapSet[T]` interface should be considered breaking changes. I'm open to feedback on a better way to structure this. I considered putting `MapSet` in `types`, but thought that might be confusing...especially because I have some desire to extract an `ImmutableHashSet` out of `types.Set` in the future.

### CODE REVIEW

File: types/entities.go [hunk 1]
```diff
@@ -13,13 +13,6 @@ import (
 // the Entity (it must be the same as the UID within the Entity itself.)
 type Entities map[EntityUID]*Entity
 
-// An Entity defines the parents and attributes for an EntityUID.
```
Comment (patjakdev) [inline types/entities.go:16]: I extracted this because I always have trouble remembering where the heck this struct is defined.

File: authorize_test.go [hunk 2]
```diff
@@ -1,68 +1,68 @@
-package cedar
+package cedar_test
```
Comment (patjakdev) [inline authorize_test.go:1]: This is good hygiene. None of the tests in this file need to interact with any private bits in the package.

File: authorize_test.go [hunk 3]
```diff
@@ -71,435 +71,435 @@ func TestIsAuthorized(t *testing.T) {
...
 		{
 			Name:   "permit-requires-entities-parent-success",
 			Policy: `permit(principal,action,resource) when { principal in parent::"bob" };`,
-			Entities: Entities{
-				cuzco: &Entity{
+			Entities: cedar.Entities{
+				cuzco: &cedar.Entity{
 					UID:     cuzco,
-					Parents: []EntityUID{types.NewEntityUID("parent", "bob")},
+					Parents: cedar.NewEntityUIDSetFromSlice([]cedar.EntityUID{cedar.NewEntityUID("parent", "bob")}),
```
Comment (patjakdev) [inline authorize_test.go:142]: Here's an example of the breaking change.

I'll admit that it's a bit of a mouthful...

File: authorize_test.go [hunk 4]
```diff
@@ -71,435 +71,435 @@ func TestIsAuthorized(t *testing.T) {
...
 		{
 			Name:   "permit-principal-in",
 			Policy: `permit(principal in team::"osiris",action,resource);`,
-			Entities: Entities{
-				cuzco: &Entity{
+			Entities: cedar.Entities{
+				cuzco: &cedar.Entity{
 					UID:     cuzco,
-					Parents: []EntityUID{types.NewEntityUID("team", "osiris")},
+					Parents: cedar.NewEntityUIDSetFromSlice([]cedar.EntityUID{cedar.NewEntityUID("team", "osiris")}),
```
Comment (patjakdev) [inline authorize_test.go:169]: Another example.

File: authorize_test.go [hunk 5]
```diff
@@ -71,435 +71,435 @@ func TestIsAuthorized(t *testing.T) {
...
 		{
 			Name:   "permit-action-in",
 			Policy: `permit(principal,action in scary::"stuff",resource);`,
-			Entities: Entities{
-				dropTable: &Entity{
+			Entities: cedar.Entities{
+				dropTable: &cedar.Entity{
 					UID:     dropTable,
-					Parents: []EntityUID{types.NewEntityUID("scary", "stuff")},
+					Parents: cedar.NewEntityUIDSetFromSlice([]cedar.EntityUID{cedar.NewEntityUID("scary", "stuff")}),
```
Comment (patjakdev) [inline authorize_test.go:196]: A couple more examples here and on line 212.

File: authorize_test.go [hunk 6]
```diff
@@ -600,183 +600,183 @@ func TestIsAuthorized(t *testing.T) {
...
 			Name:   "when-is-in",
 			Policy: `permit(principal,action,resource) when { resource is Resource in Parent::"id" };`,
-			Entities: Entities{
-				NewEntityUID("Resource", "table"): &Entity{
-					UID:     NewEntityUID("Resource", "table"),
-					Parents: []EntityUID{types.NewEntityUID("Parent", "id")},
+			Entities: cedar.Entities{
+				cedar.NewEntityUID("Resource", "table"): &cedar.Entity{
+					UID:     cedar.NewEntityUID("Resource", "table"),
+					Parents: cedar.NewEntityUIDSetFromSlice([]cedar.EntityUID{cedar.NewEntityUID("Parent", "id")}),
```
Comment (patjakdev) [inline authorize_test.go:757]: The last example in this file

File: authorize_test.go (hunk 3 above)
Comment (philhassey) [reply authorize_test.go:142]: Maybe receive a variadic, to cut out the `[]cedar.EntityUID{` ?  Though TBH, may not matter too much since it's going to be usually handled via JSON marshalling.

Maybe the words `FromSlice` aren't really needed.

File: internal/eval/evalers.go [hunk 7]
```diff
@@ -987,31 +983,28 @@ func entityInOne(env *Env, entity types.EntityUID, parent types.EntityUID) bool
...
+				if !ok || p.Parents.Len() == 0 || k == entity || known.Contains(k) {
+					return true
 				}
 				todo = append(todo, k)
-				if known == nil {
-					known = map[types.EntityUID]struct{}{}
-				}
-				known[k] = struct{}{}
-			}
+				known.Add(k)
```
Comment (philhassey) [inline internal/eval/evalers.go:1005]: I like the automatic deferred allocation.

Comment (philhassey) [reply internal/eval/evalers.go:1005]: Makes this whole bit of code more readable?

File: internal/eval/evalers_test.go [hunk 8]
```diff
@@ -1860,10 +1861,10 @@ func TestEntityIn(t *testing.T) {
 
 		entityMap := types.Entities{}
 		for i := 0; i < 100; i++ {
-			p := []types.EntityUID{
+			p := sets.NewMapSetFromSlice([]types.EntityUID{
```
Comment (philhassey) [inline internal/eval/evalers_test.go:1864]: Maybe if the package was called `mapset` you could just do `mapset.New(variadic...)`

File: internal/sets/mapset.go [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+
+// MapSet is a struct that adds some convenience to the otherwise cumbersome map[T]struct{} idiom used in Go to
+// implement sets of comparable types.
+type MapSet[T comparable] struct {
+	m map[T]struct{}
+}
+
+// NewMapSet returns a MapSet ready for use. Optionally, a desired size for the MapSet can be passed as an argument,
+// as in the argument to make() for a map type.
+func NewMapSet[T comparable](args ...int) MapSet[T] {
```
Comment (philhassey) [inline internal/sets/mapset.go:22]: Make this could be `MakeMapSet` since it is kind of like `make(map[],n)`

File: internal/sets/mapset.go [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+	var m map[T]struct{}
+	if size > 0 {
+		m = make(map[T]struct{}, size)
+	}
+
+	return MapSet[T]{m: m}
+}
+
+// NewMapSetFromSlice creates a MapSet of size len(items) and calls AddSlice(items) on it.
+func NewMapSetFromSlice[T comparable](items []T) MapSet[T] {
```
Comment (philhassey) [inline internal/sets/mapset.go:41]: And maybe this could just be `New(variadic...)`

File: internal/sets/mapset.go (hunk 9 above)
Comment (philhassey) [reply internal/sets/mapset.go:22]: or just `Make` if we call the package `mapset`

File: internal/sets/mapset.go [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+}
+
+// Add an item to the set. Returns true if the item did not exist in the set.
+func (h *MapSet[T]) Add(item T) bool {
+	if h.m == nil {
+		h.m = map[T]struct{}{}
+	}
+
+	_, exists := h.m[item]
+	h.m[item] = peppercorn
```
Comment (philhassey) [inline internal/sets/mapset.go:54]: Maybe do an `if` around the exists and just return true.  Then return false in the other case.

File: internal/sets/mapset.go [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+	return !exists
+}
+
+// AddSlice adds a slice of items to the set, returning true if any new items were added to the set.
+func (h *MapSet[T]) AddSlice(items []T) bool {
```
Comment (philhassey) [inline internal/sets/mapset.go:59]: Call this `Append` maybe?

Comment (philhassey) [reply internal/sets/mapset.go:59]: And be variadic?

File: internal/sets/mapset.go [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+	modified := false
+	for _, i := range items {
+		modified = h.Add(i) || modified
+	}
+	return modified
+}
+
+// Remove an item from the Set. Returns true if the item existed in the set.
+func (h *MapSet[T]) Remove(item T) bool {
```
Comment (philhassey) [inline internal/sets/mapset.go:68]: Could just have the one Remove and have it be variadic?

File: internal/sets/mapset.go [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+}
+
+// Contains returns whether the item exists in the set
+func (h MapSet[T]) Contains(item T) bool {
+	_, exists := h.m[item]
+	return exists
+}
+
+// Intersection returns the items common to both h and o.
+func (h MapSet[T]) Intersection(o MapSet[T]) MapSet[T] {
```
Comment (philhassey) [inline internal/sets/mapset.go:90]: This is interesting, when do we use it?

Comment (philhassey) [reply internal/sets/mapset.go:90]: The one place we use it, we just care about the Len.  So maybe instead of paying the cost of allocation, have this function instead be

`Intersects(o) bool` then it can quit early and not allocate.

File: types.go [hunk 15]
```diff
@@ -85,6 +86,17 @@ func NewEntityUID(typ EntityType, id String) EntityUID {
 	return types.NewEntityUID(typ, id)
 }
 
+// NewEntityUIDSet returns an EntityUIDSet ready for use. Optionally, a desired size for the set can be passed as an
```
Comment (philhassey) [inline types.go:89]: Not sure how useful this function is.

Comment (philhassey) [reply types.go:89]: Seems it is really only used in tests.  I'd just stick with the second, make it variadic.  Get rid of `FromSlice`

File: types/entity_uid.go [hunk 16]
```diff
@@ -91,3 +93,16 @@ func (i ImplicitlyMarshaledEntityUID) MarshalJSON() ([]byte, error) {
...
+// NewEntityUIDSet returns an EntityUIDSet ready for use. Optionally, a desired size for the EntityUIDSet can be passed
+// as an argument, as in the argument to make() for a map type.
+func NewEntityUIDSet(args ...int) EntityUIDSet {
+	return sets.NewMapSet[EntityUID](args...)
+}
+
+// NewEntityUIDSetFromSlice creates a EntityUIDSet of size len(items) and calls AddSlice(items) on it.
+func NewEntityUIDSetFromSlice(items []EntityUID) EntityUIDSet {
+	return sets.NewMapSetFromSlice[EntityUID](items)
+}
```
Comment (philhassey) [inline types/entity_uid.go:108]: Same thoughts on the naming, variadic stuff.

File: internal/parser/cedar_unmarshal.go [hunk 17]
```diff
@@ -139,11 +140,11 @@ func (p *parser) annotations() (ast.Annotations, error) {
 
 }
 
-func (p *parser) annotation(a *ast.Annotations, known map[string]struct{}) error {
+func (p *parser) annotation(a *ast.Annotations, known *sets.MapSet[string]) error {
```
Comment (philhassey) [inline internal/parser/cedar_unmarshal.go:143]: Probably doesn't need to be a pointer since it contains the map.

Comment (philhassey) [reply internal/parser/cedar_unmarshal.go:143]: Though, I suppose if the map was nil, it would need to be a pointer.

File: x/exp/batch/batch.go [hunk 18]
```diff
@@ -375,11 +382,11 @@ func cloneSub(r types.Value, k types.String, v types.Value) (types.Value, bool)
 	return r, false
 }
 
-func findVariables(found map[types.String]struct{}, r types.Value) {
+func findVariables(found *sets.MapSet[types.String], r types.Value) {
```
Comment (philhassey) [inline x/exp/batch/batch.go:385]: I wonder if sets.NewMapSet()... should return a pointer since the Add method leans on the ability to be able to modify things?

Comment (philhassey) [reply x/exp/batch/batch.go:385]: Annoyingly though that costs an allocation in `in` lookups.  So I guess it's a funny tradeoff.  Hmm.

**Review by philhassey — APPROVED** [review]
A lot of little nitpicks here, glad to discuss.  Great improvement to a lot of stuff!

File: internal/sets/mapset.go (hunk 9 above)
Comment (patjakdev) [reply internal/sets/mapset.go:22]: Groovy, if you feel like that's a better name I'm all for it.

**Comment by philhassey** [comment]
> Set has no ordering. This applies to MarshalJSON as well, meaning there's no predictable JSON encoding for a given Set. Instead of using comparable, we could invent an Item interface that embeds comparable and also has a Less() method, but then this set couldn't easily be used for native types like string, which feels like a pretty lame drawback.

I think having a consistent JSON marshaller is important because having things change randomly is really annoying and can cause unnecessary updates happen when a diff is detected by something somewhere.

In the case of JSON marshaling, I don't think performance is the key metric.  So I think what you could do is get a Marshal of the bytes of all the values.  Then sort those slices of bytes, then emit the "[a,b,c]" in that order.

>While MapSet is implemented in internal and therefore can't be directly imported by external users, the Entity.Parents is now of type MapSet[EntityUID] and therefore changes to the MapSet[T] interface should be considered breaking changes. I'm open to feedback on a better way to structure this. I considered putting MapSet in types, but thought that might be confusing...especially because I have some desire to extract an ImmutableHashSet out of types.Set in the future.

I thought the Entity.Parents was going to be immutable?

I agree that having `type X internal.Something` isn't great, but I think we can defer fixing that for a while.  We have similar problems in our public `ast` package.

File: internal/eval/evalers.go [hunk 19]
```diff
@@ -973,11 +974,6 @@ func newInEval(lhs, rhs Evaler) Evaler {
 	return &inEval{lhs: lhs, rhs: rhs}
 }
 
-func hasKnown(known map[types.EntityUID]struct{}, k types.EntityUID) bool {
-	_, ok := known[k]
-	return ok
-}
-
 func entityInOne(env *Env, entity types.EntityUID, parent types.EntityUID) bool {
 	key := inKey{a: entity, b: parent}
```
Comment (philhassey) [inline internal/eval/evalers.go:978]: I thought the benchmarks indicated the cache wasn't providing much value?

Fine to leave changing the cache to another PR though.

File: internal/eval/evalers.go [hunk 20]
```diff
@@ -1020,31 +1013,27 @@ func entityInOneWork(env *Env, entity types.EntityUID, parent types.EntityUID) b
...
 		if fe, ok := env.Entities[candidate]; ok {
-			for _, k := range fe.Parents {
-				if _, ok := parents[k]; ok {
-					return true
-				}
+			if fe.Parents.Intersects(parents) {
+				return true
 			}
-			for _, k := range fe.Parents {
+			fe.Parents.Iterate(func(k types.EntityUID) bool {
```
Comment (philhassey) [inline internal/eval/evalers.go:1028]: This is fine, but as a point of interest, I wonder if using a closure causes an allocation.

File: internal/mapset/immutable_test.go [hunk 21]
```diff
@@ -0,0 +1,166 @@
...
+		testutil.OK(t, err)
+		testutil.Equals(t, string(out), "[]")
+	})
+
+	t.Run("encode json", func(t *testing.T) {
+		s := Immutable(1, 2, 3)
+
+		out, err := json.Marshal(s)
+
+		correctOutputs := []string{
```
Comment (philhassey) [inline internal/mapset/immutable_test.go:124]: See other comment, let's not allow this.

File: internal/mapset/immutable_test.go [hunk 22]
```diff
@@ -0,0 +1,166 @@
...
+	})
+
+	t.Run("two elements", func(t *testing.T) {
+		s := Immutable[int](1, 2)
+		testutil.Equals(t, s.Contains(1), true)
+		testutil.Equals(t, s.Contains(2), true)
+		testutil.Equals(t, s.Contains(3), false)
+	})
+
+	t.Run("deduplicate elements", func(t *testing.T) {
```
Comment (philhassey) [inline internal/mapset/immutable_test.go:34]: If there is a .Len() method this might be a better place to prove deduplication.

File: internal/mapset/mapset.go [hunk 23]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+// Add an item to the set. Returns true if the item did not exist in the set.
+func (h *MapSet[T]) Add(item T) bool {
+	if h.m == nil {
+		h.m = map[T]struct{}{}
+	}
+
+	_, exists := h.m[item]
+	h.m[item] = peppercorn
```
Comment (philhassey) [inline internal/mapset/mapset.go:57]: Can you `if _,exists := { return }`

File: internal/mapset/mapset.go [hunk 24]
```diff
@@ -0,0 +1,141 @@
...
+	for item := range h.m {
+		if !o.Contains(item) {
+			return false
+		}
+	}
+	return true
+}
+
+// MarshalJSON serializes a MapSet as a JSON array. Order is non-deterministic.
+func (h MapSet[T]) MarshalJSON() ([]byte, error) {
```
Comment (philhassey) [inline internal/mapset/mapset.go:125]: Also update the comment when it becomes deterministic.

File: internal/mapset/mapset_test.go [hunk 25]
```diff
@@ -0,0 +1,234 @@
...
+
+func TestHashSet(t *testing.T) {
+	t.Run("empty set contains nothing", func(t *testing.T) {
+		s := Make[int]()
+		testutil.Equals(t, s.Len(), 0)
+		hashSetMustNotContain(t, s, 1)
+
+		s = Make[int](10)
+		testutil.Equals(t, s.Len(), 0)
+		hashSetMustNotContain(t, s, 1)
```
Comment (philhassey) [inline internal/mapset/mapset_test.go:23]: Maybe checking for contains `0` would be more interesting, since it's the zero value.

File: internal/mapset/mapset_test.go [hunk 26]
```diff
@@ -0,0 +1,234 @@
...
+
+		s2.Add(4)
+		testutil.Equals(t, s1.Equal(s2), false)
+
+		s2.Remove(3)
+		testutil.Equals(t, s1.Equal(s2), false)
+
+		s1.Add(4)
+		s1.Remove(3)
+		testutil.Equals(t, s1.Equal(s2), true)
```
Comment (philhassey) [inline internal/mapset/mapset_test.go:93]: Maye drop in a `s2.Equal(s1)`

File: policy.go [hunk 27]
```diff
@@ -97,7 +97,7 @@ func (p *Policy) Position() Position {
 	return Position(p.ast.Position)
 }
 
-// SetFilename sets the filename of this policy.
+// SetFilename mapset the filename of this policy.
```
Comment (philhassey) [inline policy.go:100]: `s/sets/mapset` too hard

File: types/entity.go [hunk 28]
```diff
@@ -0,0 +1,31 @@
...
+	UID        EntityUID    `json:"uid"`
+	Parents    EntityUIDSet `json:"parents"`
+	Attributes Record       `json:"attrs"`
+}
+
+// MarshalJSON serializes Entity as a JSON object, using the implicit form of EntityUID encoding to match the Rust
+// SDK's behavior.
+func (e Entity) MarshalJSON() ([]byte, error) {
+	parents := make([]ImplicitlyMarshaledEntityUID, 0, e.Parents.Len())
+	e.Parents.Iterate(func(p EntityUID) bool {
```
Comment (philhassey) [inline types/entity.go:16]: Throw a slices.Sort on the parents slice to keep this consistent.

Comment (philhassey) [reply types/entity.go:16]: Maybe add a test to prove the point.

File: types/set.go [hunk 29]
```diff
@@ -89,7 +89,7 @@ func (s Set) Slice() []Value {
 	return maps.Values(s.s)
 }
 
-// Equal returns true if the sets are Equal.
+// Equal returns true if the mapset are Equal.
```
Comment (philhassey) [inline types/set.go:92]: s/sets/mapset too hard

File: internal/eval/evalers.go (hunk 19 above)
Comment (patjakdev) [reply internal/eval/evalers.go:978]: Yeah, I was going to do that in a separate PR.

**Review by philhassey — APPROVED** [review]
Looks good, a few minor tweaks would be nice.
